pkg/appauth/manager/json/json.go

```diff
@@ -17,3 +17,185 @@
+)
+
+func init() {
+	registry.Register("json", New)
+}
+
+type config struct {
+	File          string `mapstructure:"file"`
+	TokenStrength int    `mapstructure:"token_strength"`
+}
+
+type jsonManager struct {
+	config *config
+	model  *jsonModel
+}
+
+type jsonModel struct {
+	file  string
+	State map[string]*apppb.AppPassword `json:"state"`
+}
+
+// New returns a new mgr.
+func New(m map[string]interface{}) (appauth.Manager, error) {
+	c, err := parseConfig(m)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating a new manager")
+	}
+
+	c.init()
+
+	// load or create file
+	model, err := loadOrCreate(c.File)
+	if err != nil {
+		return nil, errtypes.InternalError("error loading the file containing the application passwords")
+	}
+
+	return &jsonManager{config: c, model: model}, nil
+}
+
+func (c *config) init() {
+	if c.File == "" {
+		c.File = "/etc/revad/appauth.json"
+	}
+}
+
+func parseConfig(m map[string]interface{}) (*config, error) {
+	c := &config{}
+	if err := mapstructure.Decode(m, c); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func loadOrCreate(file string) (*jsonModel, error) {
+	stat, err := os.Stat(file)
+	if os.IsNotExist(err) || stat.Size() == 0 {
+		if err = ioutil.WriteFile(file, []byte("{}"), 0644); err != nil {
+			return nil, errors.Wrapf(err, "error creating the file %s", file)
+		}
+	}
+
+	fd, err := os.OpenFile(file, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error opening the file %s", file)
+	}
+	defer fd.Close()
+
+	data, err := ioutil.ReadAll(fd)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading the file %s", file)
+	}
+
+	m := &jsonModel{file: file}
+	if err = json.Unmarshal(data, &m.State); err != nil {
+		return nil, errors.Wrapf(err, "error parsing the file %s", file)
+	}
+
+	if m.State == nil {
+		m.State = make(map[string]*apppb.AppPassword)
+	}
+	return m, nil
+}
+
+func (mgr *jsonManager) GenerateAppPassword(ctx context.Context, scope map[string]*authpb.Scope, label string, expiration *typespb.Timestamp) (*apppb.AppPassword, error) {
+	token, err := password.Generate(mgr.config.TokenStrength, 10, 10, false, false)
+	if err != nil {
+		return nil, errtypes.InternalError("error creating new token")
+	}
+	user := user.ContextMustGetUser(ctx)
+	ctime := now()
+
+	appPass := &apppb.AppPassword{
+		Password:   token,
+		TokenScope: scope,
+		Label:      label,
+		Expiration: expiration,
+		Ctime:      ctime,
+		Utime:      ctime,
+		User:       user.GetId(),
+	}
+	mgr.model.State[token] = appPass
+
+	err = mgr.model.save()
+	if err != nil {
+		return nil, errtypes.InternalError("error saving new token")
+	}
+
+	return appPass, nil
+}
+
+func (mgr *jsonManager) ListAppPasswords(ctx context.Context) ([]*apppb.AppPassword, error) {
+	userId := user.ContextMustGetUser(ctx).GetId()
+	userPasswords := []*apppb.AppPassword{}
+
+	for _, appPassword := range mgr.model.State {
+		if reflect.DeepEqual(appPassword.User, userId) {
```

Comment:
I personally don't like using DeepEqual, because it's pretty fragile. We anyway have a function in the utils (`utils.UserEqual`) pkg, so we can use that.